File: src/tools/RenamerFv.cpp
#include "RenamerFv.h"
#include "../SecurityManager.h"
#include "../Utils.h"
#include <QApplication>
#include <QDesktopServices>
#include <QDir>
#include <QFileDialog>
#include <QHBoxLayout>
#include <QLabel>
#include <QMessageBox>
#include <QRegularExpression>
#include <QUrl>
#include <QVBoxLayout>
#include <poppler/cpp/poppler-document.h>
#include <poppler/cpp/poppler-page.h>

namespace GOL {

RenamerFv::RenamerFv(QWidget *parent) : QDialog(parent) {
  // Security Check
  SecurityManager::instance().checkAndAct();

  setWindowTitle("GOLEVENTS - TICKET RENAMER PRO 🏷️");
  resize(900, 750);
  setStyleSheet(
      QString("background-color: %1; color: white;").arg(Utils::BG_COLOR));

  QVBoxLayout *mainLayout = new QVBoxLayout(this);
  mainLayout->setContentsMargins(40, 30, 40, 40);
  mainLayout->setSpacing(20);

  // Header
  QLabel *title = new QLabel("🏷️ TICKET RENAMER + FV");
  title->setStyleSheet(QString("font-size: 30px; font-weight: bold; color: %1;")
                           .arg(Utils::ACCENT_COLOR));
  mainLayout->addWidget(title);

  QLabel *sub = new QLabel(
      "Auto-detect prices inside PDFs and rename files using strict format.");
  sub->setStyleSheet("font-size: 13px; color: gray;");
  mainLayout->addWidget(sub);

  // Folder selection
  QHBoxLayout *pathLayout = new QHBoxLayout();
  m_pathEdit = new QLineEdit();
  m_pathEdit->setPlaceholderText("Click Browse to select folder...");
  m_pathEdit->setReadOnly(true);
  m_pathEdit->setFixedHeight(45);
  m_pathEdit->setStyleSheet(
      QString("background-color: #0d0d0d; border: 1px solid %1; border-radius: "
              "10px; padding: 0 15px;")
          .arg(Utils::CARD_BORDER));

  m_btnBrowse = new QPushButton("BROWSE");
  m_btnBrowse->setFixedSize(120, 45);
  m_btnBrowse->setStyleSheet(QString("background-color: %1; color: white; "
                                     "border-radius: 10px; font-weight: bold;")
                                 .arg(Utils::ACCENT_COLOR));

  pathLayout->addWidget(m_pathEdit);
  pathLayout->addWidget(m_btnBrowse);
  mainLayout->addLayout(pathLayout);

  // Run button
  m_btnRun = new QPushButton("🚀 START BULK RENAMING");
  m_btnRun->setFixedHeight(60);
  m_btnRun->setStyleSheet(
      "background-color: #10b981; color: black; border-radius: 15px; "
      "font-weight: bold; font-size: 16px;");
  mainLayout->addWidget(m_btnRun);

  // Log area
  m_logArea = new QTextEdit();
  m_logArea->setReadOnly(true);
  m_logArea->setFont(QFont("Consolas", 12));
  m_logArea->setStyleSheet(
      QString("background-color: #050505; color: #00FF88; border: 1px solid "
              "%1; border-radius: 10px; padding: 10px;")
          .arg(Utils::CARD_BORDER));
  mainLayout->addWidget(m_logArea);

  // Connections
  connect(m_btnBrowse, &QPushButton::clicked, this, &RenamerFv::browseFolder);
  connect(m_btnRun, &QPushButton::clicked, this, &RenamerFv::startProcess);
}

void RenamerFv::log(const QString &msg) { m_logArea->append("> " + msg); }

void RenamerFv::browseFolder() {
  QString p = QFileDialog::getExistingDirectory(this, "Select folder");
  if (!p.isEmpty())
    m_pathEdit->setText(p);
}

QString RenamerFv::extractFaceValue(const QString &pdfPath) {
  try {
    poppler::document *doc =
        poppler::document::load_from_file(pdfPath.toStdString());
    if (!doc)
      return "";

    QString fullText;
    int numPages = doc->pages();
    for (int i = 0; i < numPages; ++i) {
      poppler::page *p = doc->create_page(i);
      if (p) {
        poppler::byte_array ba = p->text().to_utf8();
        fullText += QString::fromUtf8(ba.data(), ba.size()) + "\n";
        delete p;
      }
    }
    delete doc;

    if (fullText.isEmpty())
      return "";

    // Clean text: remove spaces between numbers (e.g., 1 5 0 . 0 0 -> 150.00)
    QRegularExpression spaceNumRe("(\\d)\\s*([.,])\\s*(\\d)");
    fullText.replace(spaceNumRe, "\\1\\2\\3");

    // Pattern for Prices: 1 to 4 digits before dot, exactly 2 digits after
    QRegularExpression pricePattern("\\b\\d{1,4}[.,]\\d{2}\\b");

    // 1. Target search with Keywords
    QString keywords =
        "(?:TOTALE|TOTAL|IMPORTO|VALORE|PREZZO|PRICE|PRIX|PRECIO)";
    QRegularExpression searchPattern(
        keywords +
            "\\s*(?:TICKET|AMOUNT|EUR|€)?[:\\s€]*(\\b\\d{1,4}[.,]\\d{2}\\b)",
        QRegularExpression::CaseInsensitiveOption);

    auto it = searchPattern.globalMatch(fullText);
    QString lastMatch;
    while (it.hasNext()) {
      lastMatch = it.next().captured(1);
    }

    if (!lastMatch.isEmpty()) {
      return lastMatch.replace(",", ".");
    }

    // 2. Fallback: Get the HIGHEST number matching the format
    auto itAll = pricePattern.globalMatch(fullText);
    double maxP = -1.0;
    while (itAll.hasNext()) {
      QString cand = itAll.next().captured(0).replace(",", ".");
      double val = cand.toDouble();
      if (val > maxP)
        maxP = val;
    }

    if (maxP >= 0) {
      return QString::number(maxP, 'f', 2);
    }

    return "";
  } catch (...) {
    return "";
  }
}

void RenamerFv::startProcess() {
  // Security Check
  SecurityManager::instance().checkAndAct();

  QString folder = m_pathEdit->text();
  if (folder.isEmpty() || !QDir(folder).exists()) {
    QMessageBox::warning(this, "Warning",
                         "Please select a valid folder first.");
    return;
  }

  m_btnRun->setEnabled(false);
  m_btnRun->setText("PROCESSING...");
  m_logArea->clear();
  log("Scanning folder: " + folder);

  QDir dir(folder);
  QStringList files = dir.entryList({"*.pdf"}, QDir::Files);

  int successCount = 0;
  for (const QString &filename : files) {
    if (filename.toUpper().contains("-FV"))
      continue; // Skip already processed

    QString fullPath = dir.absoluteFilePath(filename);
    QString price = extractFaceValue(fullPath);

    if (!price.isEmpty()) {
      QString cleanPrice = price;
      cleanPrice.replace(".", "p");

      QString baseName = QFileInfo(filename).baseName();
      QString newFilename = baseName + "-FV" + cleanPrice + ".pdf";

      if (dir.rename(filename, newFilename)) {
        log(QString("SUCCESS: %1 ⮕ FV %2€").arg(filename).arg(price));
        successCount++;
      } else {
        log(QString("ERROR: Could not rename %1").arg(filename));
      }
    } else {
      log(QString("SKIPPED: Format XX.YY not found in %1").arg(filename));
    }

    QApplication::processEvents(); // Keep UI responsive
  }

  log(QString("\n✅ FINISHED! Total %1 tickets renamed.").arg(successCount));
  m_btnRun->setEnabled(true);
  m_btnRun->setText("🚀 START BULK RENAMING");

  QMessageBox::information(
      this, "Success",
      QString("Renaming Complete!\nProcessed %1 files.").arg(successCount));
  QDesktopServices::openUrl(QUrl::fromLocalFile(folder));
}

} // namespace GOL

File: legacy_python/scripts/renamer_fv.py
import os
import re
import threading
import customtkinter as ctk
import requests
import subprocess
import sys
from tkinter import messagebox, filedialog
from PyPDF2 import PdfReader

# ================== [ SECURITY & DESIGN CONFIG ] ==================
VERSION = "2.9"
FIREBASE_URL = "https://golevents1-default-rtdb.europe-west1.firebasedatabase.app"
ACCENT_COLOR = "#1f6aa5" 
PRO_COLOR = "#ffcc00"
BG_COLOR = "#0b0b0b"
CARD_BG = "#161616"
CARD_BORDER = "#252525"
# =================================================================

def get_hwid():
    try:
        cmd = 'wmic baseboard get serialnumber'
        hwid = str(subprocess.check_output(cmd, shell=True), 'utf-8').split('\n')[1].strip()
        return hwid
    except:
        import hashlib
        return hashlib.sha256(os.environ['COMPUTERNAME'].encode()).hexdigest()[:16]

# --- SECURITY WRAPPER ---
class SecurityCheck(ctk.CTk):
    def __init__(self, key):
        super().__init__()
        self.withdraw()
        self.license_key = key
        
        self.splash = ctk.CTkToplevel(self)
        self.splash.title("Authenticating...")
        self.splash.geometry("450x250")
        self.splash.configure(fg_color=BG_COLOR)
        self.splash.overrideredirect(True)
        
        x = (self.winfo_screenwidth() // 2) - 225
        y = (self.winfo_screenheight() // 2) - 125
        self.splash.geometry(f"+{x}+{y}")

        ctk.CTkLabel(self.splash, text="GOLEVENTS SECURE 👑", font=("Arial", 24, "bold"), text_color=ACCENT_COLOR).pack(pady=(45, 5))
        self.status_lbl = ctk.CTkLabel(self.splash, text="Verifying Module Access...", font=("Arial", 13), text_color="gray")
        self.status_lbl.pack()
        
        self.progress = ctk.CTkProgressBar(self.splash, width=350, height=12, progress_color=ACCENT_COLOR)
        self.progress.set(0)
        self.progress.pack(pady=30)
        
        self.after(500, self.verify)

    def verify(self):
        hwid = get_hwid()
        try:
            res = requests.get(f"{FIREBASE_URL}/keys/{self.license_key}.json").json()
            if res and res.get("status") == "active" and res.get("hwid") == hwid:
                self.progress.set(1.0)
                self.status_lbl.configure(text="ACCESS GRANTED", text_color="#2ecc71")
                self.after(800, self.launch_main_tool)
            else:
                messagebox.showerror("Security Alert", "Unauthorized Access!\nPlease launch through GOLEVENTS Hub.")
                sys.exit()
        except:
            sys.exit()

    def launch_main_tool(self):
        self.splash.destroy()
        self.deiconify()
        TicketRenamerApp(self)

# --- CORE LOGIC: ADVANCED FACE VALUE EXTRACTION ---
def extract_face_value(pdf_path):
    try:
        reader = PdfReader(pdf_path)
        text = ""
        for page in reader.pages:
            page_text = page.extract_text()
            if page_text:
                # Clean text: remove spaces between numbers (e.g., 1 5 0 . 0 0 -> 150.00)
                page_text = re.sub(r'(\d)\s*([.,])\s*(\d)', r'\1\2\3', page_text)
                text += page_text + "\n"

        # Pattern for Prices: X.XX to XXXX.XX
        # Strict format: 1 to 4 digits before dot, exactly 2 digits after.
        price_pattern = r'\b\d{1,4}[.,]\d{2}\b'

        # 1. Target search with Keywords
        keywords = r'(?:TOTALE|TOTAL|IMPORTO|VALORE|PREZZO|PRICE|PRIX|PRECIO)'
        search_pattern = keywords + r'\s*(?:TICKET|AMOUNT|EUR|€)?[:\s€]*(\b\d{1,4}[.,]\d{2}\b)'
        
        matches = re.findall(search_pattern, text, re.IGNORECASE)
        if matches:
            return matches[-1].replace(",", ".") # Return the last found total

        # 2. Fallback: Get the HIGHEST number matching the format
        all_candidates = re.findall(price_pattern, text)
        if all_candidates:
            float_prices = [float(p.replace(",", ".")) for p in all_candidates]
            max_p = max(float_prices)
            return "{:.2f}".format(max_p)

        return None
    except:
        return None

# --- GUI INTERFACE ---
class TicketRenamerApp:
    def __init__(self, master):
        self.master = master
        self.master.title("GOLEVENTS - TICKET RENAMER PRO 🏷️")
        self.master.geometry("900x750")
        self.master.configure(fg_color=BG_COLOR)

        # Header
        header_frame = ctk.CTkFrame(self.master, fg_color="transparent")
        header_frame.pack(fill="x", padx=40, pady=(30, 10))
        
        ctk.CTkLabel(header_frame, text="🏷️ TICKET RENAMER + FV", font=("Arial", 30, "bold"), text_color=ACCENT_COLOR).pack(anchor="w")
        ctk.CTkLabel(header_frame, text="Auto-detect prices inside PDFs and rename files using strict XX.YY format.", font=("Arial", 13), text_color="gray60").pack(anchor="w")

        # Folder Selection
        self.path_var = ctk.StringVar(value="Click Browse to select folder...")
        path_frame = ctk.CTkFrame(self.master, fg_color=CARD_BG, corner_radius=15, border_width=1, border_color=CARD_BORDER)
        path_frame.pack(fill="x", padx=40, pady=20)

        ctk.CTkEntry(path_frame, textvariable=self.path_var, width=600, height=45, state="readonly", fg_color="#0d0d0d").pack(side="left", padx=20, pady=20)
        ctk.CTkButton(path_frame, text="BROWSE", width=120, height=45, fg_color=ACCENT_COLOR, font=("Arial", 12, "bold"), command=self.browse).pack(side="left", padx=(0, 20))

        # Action Button
        self.run_btn = ctk.CTkButton(self.master, text="🚀 START BULK RENAMING", height=60, 
                                      fg_color="#10b981", hover_color="#059669", text_color="black",
                                      font=("Arial", 16, "bold"), command=self.start_process)
        self.run_btn.pack(pady=10, padx=40, fill="x")

        # Logs
        self.log_area = ctk.CTkTextbox(self.master, height=250, fg_color="#050505", text_color="#00FF88", 
                                        font=("Consolas", 12), border_width=1, border_color=CARD_BORDER)
        self.log_area.pack(fill="both", expand=True, padx=40, pady=20)

    def browse(self):
        p = filedialog.askdirectory()
        if p: self.path_var.set(p)

    def log(self, msg):
        self.log_area.configure(state="normal")
        self.log_area.insert("end", f"> {msg}\n")
        self.log_area.see("end")
        self.log_area.configure(state="disabled")
        self.master.update()

    def start_process(self):
        folder = self.path_var.get()
        if folder == "Click Browse to select folder..." or not os.path.isdir(folder):
            messagebox.showwarning("Warning", "Please select a valid folder first.")
            return

        self.run_btn.configure(state="disabled", text="PROCESSING...")
        threading.Thread(target=self.process_files, args=(folder,), daemon=True).start()

    def process_files(self, folder):
        self.log(f"Scanning folder: {folder}")
        files = [f for f in os.listdir(folder) if f.lower().endswith(".pdf") and "-FV" not in f.upper()]
        
        if not files:
            self.log("No new PDF files found or all files already have FV tags.")
            self.run_btn.configure(state="normal", text="🚀 START BULK RENAMING")
            return

        success_count = 0
        for filename in files:
            old_path = os.path.join(folder, filename)
            price = extract_face_value(old_path)
            
            if price:
                # Format: replace dot with 'p' (e.g., 50.00 -> 50p00)
                clean_price = price.replace(".", "p")
                base_name = os.path.splitext(filename)[0]
                new_filename = f"{base_name}-FV{clean_price}.pdf"
                new_path = os.path.join(folder, new_filename)
                
                try:
                    if not os.path.exists(new_path):
                        os.rename(old_path, new_path)
                        self.log(f"SUCCESS: {filename} ⮕ FV {price}€")
                        success_count += 1
                except Exception as e:
                    self.log(f"ERROR: {filename} | {e}")
            else:
                self.log(f"SKIPPED: Format XX.YY not found in {filename}")

        self.log(f"\n✅ FINISHED! Total {success_count} tickets renamed.")
        self.run_btn.configure(state="normal", text="🚀 START BULK RENAMING")
        messagebox.showinfo("Success", f"Renaming Complete!\nProcessed {success_count} files.")
        try: os.startfile(folder)
        except: pass

if __name__ == "__main__":
    if len(sys.argv) > 1:
        app = SecurityCheck(sys.argv[1])
        app.mainloop()
    else:
        ctk.set_appearance_mode("Dark")
        root = ctk.CTk(); root.withdraw()
        messagebox.showerror("Error", "Launch via GOLEVENTS Hub.")
        sys.exit()